sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala

```diff
@@ -131,20 +244,70 @@ object SchemaUtil {
     val stateVarType = stateVarInfo.stateVariableType
 
     stateVarType match {
-      case StateVariableType.ValueState =>
+      case ValueState =>
         new StructType()
           .add("key", stateStoreColFamilySchema.keySchema)
           .add("single_value", stateStoreColFamilySchema.valueSchema)
           .add("partition_id", IntegerType)
 
-      case StateVariableType.ListState =>
+      case ListState =>
         new StructType()
           .add("key", stateStoreColFamilySchema.keySchema)
           .add("list_value", ArrayType(stateStoreColFamilySchema.valueSchema))
           .add("partition_id", IntegerType)
 
+      case MapState =>
+        val groupingKeySchema = SchemaUtil.getSchemaAsDataType(
+          stateStoreColFamilySchema.keySchema, "key")
+        val userKeySchema = stateStoreColFamilySchema.userKeyEncoderSchema.get
+        val valueMapSchema = MapType.apply(
+          keyType = userKeySchema,
+          valueType = stateStoreColFamilySchema.valueSchema
+        )
+
+        new StructType()
+          .add("key", groupingKeySchema)
+          .add("map_value", valueMapSchema)
+          .add("partition_id", IntegerType)
+
       case _ =>
         throw StateDataSourceErrors.internalError(s"Unsupported state variable type $stateVarType")
     }
   }
+
+  /**
+   * Helper functions for map state data source reader.
+   *
+   * Map state variables are stored in RocksDB state store has the schema of
+   * `TransformWithStateKeyValueRowSchemaUtils.getCompositeKeySchema()`;
+   * But for state store reader, we need to return in format of:
+   * "key": groupingKey, "map_value": Map(userKey -> value).
+   *
+   * The following functions help to translate between two schema.
+   */
+  def isMapStateVariable(
+      stateVariableInfoOpt: Option[TransformWithStateVariableInfo]): Boolean = {
+    stateVariableInfoOpt.isDefined &&
+      stateVariableInfoOpt.get.stateVariableType == MapState
+  }
+
+  /**
+   * Given key-value schema generated from `generateSchemaForStateVar()`,
+   * returns the compositeKey schema that key is stored in the state store
+   */
+  def getCompositeKeySchema(schema: StructType): StructType = {
+    val groupingKeySchema = SchemaUtil.getSchemaAsDataType(
+      schema, "key").asInstanceOf[StructType]
+    val userKeySchema = try {
+      Option(
+        SchemaUtil.getSchemaAsDataType(schema, "map_value").asInstanceOf[MapType]
+          .keyType.asInstanceOf[StructType])
+    } catch {
+      case _: Exception =>
+        None
+    }
+    new StructType()
+      .add("key", groupingKeySchema)
+      .add("userKey", userKeySchema.get)
```

Comment:
I'd rather not leave this to raise NoSuchElementException which would not be helpful for both users and developers. Shall we explicitly throw an internal exception with message containing context of this error?